// ng-solcre-auth/esm2015/lib/auth.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { Validators, FormBuilder } from '@angular/forms';
import { AuthService } from './solcre-auth-library.service';
export class SolcreAuthLibraryComponent {
    /**
     * @param {?} formBuilder
     * @param {?} authService
     */
    constructor(formBuilder, authService) {
        this.formBuilder = formBuilder;
        this.authService = authService;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.signinForm = this.formBuilder.group({
            'email': this.formBuilder.control(null, [Validators.required, Validators.email]),
            'password': this.formBuilder.control(null, Validators.required)
        });
        this.authService.codeNotFound.subscribe((/**
         * @param {?} state
         * @return {?}
         */
        (state) => {
            this.codeNotFound = state;
        }));
        this.authService.searchingCode.subscribe((/**
         * @param {?} state
         * @return {?}
         */
        (state) => {
            this.searchingCode = state;
        }));
    }
    /**
     * @return {?}
     */
    onSubmit() {
        this.authService.login(this.signinForm.value.email, this.signinForm.value.password);
    }
    /**
     * @return {?}
     */
    onBlur() {
        this.codeNotFound = false;
        this.searchingCode = false;
        this.signinForm.invalid;
        if (this.signinForm.value.email && this.signinForm.value.email.indexOf('@') > -1) {
            /** @type {?} */
            let data = (this.signinForm.value.email).split("@", 2);
            /** @type {?} */
            let domain = data[1];
            this.authService.setCode(domain);
        }
    }
}
SolcreAuthLibraryComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-auth-library',
                template: "<form \n    name=\"app-login-form\"\n    id=\"app-login-form\"\n    class=\"styled-form\" \n    [formGroup]=\"signinForm\" \n    (ngSubmit)=\"onSubmit()\">\n    <h1>Auth library</h1>\n    <div>\n        <input class=\"input\" type=\"text\" (blur)=\"onBlur()\" id=\"email\" formControlName=\"email\" name=\"email\"><br>\n    </div>\n    <label *ngIf=\"searchingCode\" for=\"email\">Buscando dominio...</label>\n    <label *ngIf=\"codeNotFound\" for=\"email\">*Error al encontrar dominio</label>\n    <div>\n        <input class=\"input\" type=\"password\" id=\"password\" formControlName=\"password\" name=\"password\"><br>\n    </div>\n    <button class=\"rounded-btn\" type=\"submit\">Entrar</button>\n</form>"
            }] }
];
/** @nocollapse */
SolcreAuthLibraryComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: AuthService }
];
if (false) {
    /** @type {?} */
    SolcreAuthLibraryComponent.prototype.signinForm;
    /** @type {?} */
    SolcreAuthLibraryComponent.prototype.searchingCode;
    /** @type {?} */
    SolcreAuthLibraryComponent.prototype.codeNotFound;
    /**
     * @type {?}
     * @private
     */
    SolcreAuthLibraryComponent.prototype.formBuilder;
    /**
     * @type {?}
     * @private
     */
    SolcreAuthLibraryComponent.prototype.authService;
}
