// ng-solcre-auth/esm2015/lib/auth.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { Observable, BehaviorSubject, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { AuthService } from './solcre-auth-library.service';
// ​import { AccessTokenModel } from './access-token.model';
export class AuthInterceptor {
    /**
     * @param {?} authService
     */
    constructor(authService) {
        this.authService = authService;
        this.isRefreshingToken = false;
        this.tokenSubject = new BehaviorSubject(null);
    }
    //Intercep method
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        // Pass on the cloned request instead of the original request.
        return (/** @type {?} */ (next.handle(req)
            .pipe(catchError((/**
         * @param {?} error
         * @param {?} caught
         * @return {?}
         */
        (error, caught) => {
            //Check error type
            if (error instanceof HttpErrorResponse) {
                console.log(error);
                switch (error.status) {
                    case 400:
                        return this.handle400Error(error);
                    case 401:
                        return this.handle401Error(error, req, next);
                    case 403:
                        return this.handle403Error(error);
                }
            }
            return new Observable();
        })))));
    }
    //Add authorization header to requests
    /**
     * @private
     * @param {?} req
     * @param {?} token
     * @return {?}
     */
    addToken(req, token) {
        return req.clone({
            setHeaders: {
                Authorization: 'Bearer ' + token
            }
        });
    }
    //Helper function when the refresh token doesnt work
    /**
     * @private
     * @param {?} error
     * @return {?}
     */
    logout(error) {
        //logout users, redirect to login page
        this.authService.logout();
        return throwError(error);
    }
    //Handle 403 error
    /**
     * @private
     * @param {?} error
     * @return {?}
     */
    handle403Error(error) {
        return this.logout(error.message);
    }
    //Hanfle 401 error
    /**
     * @private
     * @param {?} error
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    handle401Error(error, req, next) {
        //@@TODO: find a way to configure it
        //Ignore 401 status when the url are Oauth
        if (req.url.indexOf('/oauth') > -1) {
            return throwError(error);
        }
        return this.logout(error.message);
    }
    //Handle 400 error
    /**
     * @private
     * @param {?} error
     * @return {?}
     */
    handle400Error(error) {
        if (error
            && error.status === 400
            && error.error
            && error.error.error === 'invalid_grant') {
            // If we get a 400 and the error message is 'invalid_grant', the token is no longer valid so logout.
            return this.logout(error.message);
        }
        //Normal flow
        return throwError(error);
    }
}
AuthInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AuthInterceptor.ctorParameters = () => [
    { type: AuthService }
];
if (false) {
    /** @type {?} */
    AuthInterceptor.prototype.isRefreshingToken;
    /** @type {?} */
    AuthInterceptor.prototype.tokenSubject;
    /**
     * @type {?}
     * @private
     */
    AuthInterceptor.prototype.authService;
}
